## Implementation of Simple Linear Regression From Scratch

- With the Use of
1. Python Lists
2. Aggregate Functions

- Comaprision 
1. Comparision with Sklearn implemented models
2. Accauracy Comparision

In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression

In [2]:
df = pd.read_csv('Sample_Data.csv')

In [3]:
df.head()

CGPA   LPA
0   8.5   9.2
1   8.1   8.7
2   7.9   7.5
3   8.3   8.9
4   9.2  10.5

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   CGPA    120 non-null    float64
 1   LPA     120 non-null    float64
dtypes: float64(2)
memory usage: 2.0 KB


In [5]:
output = [df.LPA.values][0]

In [6]:
print(output)

[ 9.2  8.7  7.5  8.9 10.5  7.2  9.5 10.   8.8  9.6 11.2  7.   8.4 10.2
  9.3  7.3  9.8 10.8  6.8  8.   9.2  8.7  7.5  8.9 10.5  7.2  9.5 10.
  8.8  9.6 11.2  7.   8.4 10.2  9.3  7.3  9.8 10.8  6.8  8.   9.2  8.7
  7.5  8.9 10.5  7.2  9.5 10.   8.8  9.6 11.2  7.   8.4 10.2  9.3  7.3
  9.8 10.8  6.8  8.   9.2  8.7  7.5  8.9 10.5  7.2  9.5 10.   8.8  9.6
 11.2  7.   8.4 10.2  9.3  7.3  9.8 10.8  6.8  8.   9.2  8.7  7.5  8.9
 10.5  7.2  9.5 10.   8.8  9.6 11.2  7.   8.4 10.2  9.3  7.3  9.8 10.8
  6.8  8.   9.2  8.7  7.5  8.9 10.5  7.2  9.5 10.   8.8  9.6 11.2  7.
  8.4 10.2  9.3  7.3  9.8 10.8  6.8  8. ]


In [7]:
input_ = [df.CGPA.values][0]

In [8]:
input_

array([8.5, 8.1, 7.9, 8.3, 9.2, 7.8, 8.7, 9. , 8.4, 8.8, 9.5, 7.6, 8.2,
       9.1, 8.6, 7.7, 8.9, 9.3, 7.5, 8. , 8.5, 8.1, 7.9, 8.3, 9.2, 7.8,
       8.7, 9. , 8.4, 8.8, 9.5, 7.6, 8.2, 9.1, 8.6, 7.7, 8.9, 9.3, 7.5,
       8. , 8.5, 8.1, 7.9, 8.3, 9.2, 7.8, 8.7, 9. , 8.4, 8.8, 9.5, 7.6,
       8.2, 9.1, 8.6, 7.7, 8.9, 9.3, 7.5, 8. , 8.5, 8.1, 7.9, 8.3, 9.2,
       7.8, 8.7, 9. , 8.4, 8.8, 9.5, 7.6, 8.2, 9.1, 8.6, 7.7, 8.9, 9.3,
       7.5, 8. , 8.5, 8.1, 7.9, 8.3, 9.2, 7.8, 8.7, 9. , 8.4, 8.8, 9.5,
       7.6, 8.2, 9.1, 8.6, 7.7, 8.9, 9.3, 7.5, 8. , 8.5, 8.1, 7.9, 8.3,
       9.2, 7.8, 8.7, 9. , 8.4, 8.8, 9.5, 7.6, 8.2, 9.1, 8.6, 7.7, 8.9,
       9.3, 7.5, 8. ])

In [9]:
output_mean = 0
for  i in range(len(output)):
    output_mean = output_mean + output[i]
output_mean = output_mean / len(output)

In [10]:
output_mean

8.934999999999999

In [11]:
input_mean = 0
for  i in input_:
    input_mean += i
input_mean = input_mean/len(input_)

In [12]:
y_error = []
x_error = []
for i in range(len(output)):
    y_error.append(round((output[i]-output_mean),2))
    x_error.append(round((input_[i]-input_mean),2))


In [13]:
print(y_error)

[0.27, -0.23, -1.43, -0.03, 1.57, -1.73, 0.57, 1.07, -0.13, 0.67, 2.27, -1.93, -0.53, 1.27, 0.37, -1.63, 0.87, 1.87, -2.13, -0.93, 0.27, -0.23, -1.43, -0.03, 1.57, -1.73, 0.57, 1.07, -0.13, 0.67, 2.27, -1.93, -0.53, 1.27, 0.37, -1.63, 0.87, 1.87, -2.13, -0.93, 0.27, -0.23, -1.43, -0.03, 1.57, -1.73, 0.57, 1.07, -0.13, 0.67, 2.27, -1.93, -0.53, 1.27, 0.37, -1.63, 0.87, 1.87, -2.13, -0.93, 0.27, -0.23, -1.43, -0.03, 1.57, -1.73, 0.57, 1.07, -0.13, 0.67, 2.27, -1.93, -0.53, 1.27, 0.37, -1.63, 0.87, 1.87, -2.13, -0.93, 0.27, -0.23, -1.43, -0.03, 1.57, -1.73, 0.57, 1.07, -0.13, 0.67, 2.27, -1.93, -0.53, 1.27, 0.37, -1.63, 0.87, 1.87, -2.13, -0.93, 0.27, -0.23, -1.43, -0.03, 1.57, -1.73, 0.57, 1.07, -0.13, 0.67, 2.27, -1.93, -0.53, 1.27, 0.37, -1.63, 0.87, 1.87, -2.13, -0.93]


In [14]:
print(x_error)

[0.04, -0.36, -0.56, -0.16, 0.74, -0.66, 0.24, 0.54, -0.06, 0.34, 1.04, -0.86, -0.26, 0.64, 0.14, -0.76, 0.44, 0.84, -0.96, -0.46, 0.04, -0.36, -0.56, -0.16, 0.74, -0.66, 0.24, 0.54, -0.06, 0.34, 1.04, -0.86, -0.26, 0.64, 0.14, -0.76, 0.44, 0.84, -0.96, -0.46, 0.04, -0.36, -0.56, -0.16, 0.74, -0.66, 0.24, 0.54, -0.06, 0.34, 1.04, -0.86, -0.26, 0.64, 0.14, -0.76, 0.44, 0.84, -0.96, -0.46, 0.04, -0.36, -0.56, -0.16, 0.74, -0.66, 0.24, 0.54, -0.06, 0.34, 1.04, -0.86, -0.26, 0.64, 0.14, -0.76, 0.44, 0.84, -0.96, -0.46, 0.04, -0.36, -0.56, -0.16, 0.74, -0.66, 0.24, 0.54, -0.06, 0.34, 1.04, -0.86, -0.26, 0.64, 0.14, -0.76, 0.44, 0.84, -0.96, -0.46, 0.04, -0.36, -0.56, -0.16, 0.74, -0.66, 0.24, 0.54, -0.06, 0.34, 1.04, -0.86, -0.26, 0.64, 0.14, -0.76, 0.44, 0.84, -0.96, -0.46]


In [15]:
cov_xy = 0
var_x = 0

In [16]:
for i in range(len(y_error)):
    cov_xy +=  y_error[i]*x_error[i]
    var_x += x_error[i]*x_error[i]

In [17]:
print(cov_xy,var_x,sep='\n')

89.04600000000005
41.099999999999994


In [18]:
m = cov_xy/var_x

In [19]:
print(m)

2.166569343065695


In [20]:
b = output_mean - m*input_mean

In [21]:
b

-9.383343795620455

### According to our Calculations :
- The Values of m(Wieghtage of input parameter / slope m) = 2.16
- The value of offset / intercept b = -9.38
- Equation nieng -> y = m*x + b

### Calculating parametres  with sklearn model

In [22]:
lr = LinearRegression()

In [23]:
lr.fit(df.CGPA.values.reshape(-1,1),df.LPA.values.reshape(-1,1))

LinearRegression()

In [24]:
lr.intercept_

array([-9.3852982])

In [25]:
lr.coef_

array([[2.1668005]])

#### Comparizion

- Value of m calculated manually  2.166569343065695
- Value of b calculate manually  -9.383343795620455 

----

- Value of m calculated by sklearn  2.1668005
- Value of b calculated by sklearn -9.3852982

In [26]:
test_input = np.array([7.93, 7.85, 6.23, 6.93, 6.72, 7.38, 8.69, 7.73, 8.05, 6.71, 7.59, 8.33, 9.03, 8.38, 8.22, 6.89, 9.44, 9.41, 6.56, 8.42, 8.47, 7.32, 7.77, 9.50, 7.05])
test_output = np.array([8.34, 7.22, 14.87, 11.80, 11.10, 6.05, 8.83, 5.52, 12.95, 11.85, 8.13, 8.23, 9.17, 5.85, 10.67, 7.73, 5.87, 8.90, 10.58, 8.21, 7.90, 7.27, 9.67, 5.24, 11.37])

In [27]:
print(len(test_input),len(test_output))

25 25


In [28]:
lr.predict(test_input[0].reshape(-1,1))

array([[7.79742974]])

In [29]:
test_output[0]

8.34

In [30]:
y_output = m*test_input[0] + b

In [31]:
print(y_output)

7.797551094890505


In [32]:
# Almost Same accuaracy

In [33]:
import random

In [34]:
from random import shuffle

In [35]:
df.shape[0]

120

In [36]:
a = [i for i in range(df.shape[0])]

In [37]:
random.shuffle(a)

In [38]:
df.iloc[a[0:10],:1]

CGPA
78   7.5
8    8.4
28   8.4
83   8.3
29   8.8
91   7.6
0    8.5
2    7.9
68   8.4
67   9.0

# ________________________________________________________________

## Creating Our Own Regression Class

In [39]:
class SLR:   #Simple Linear Regression
    def __init__(self):
        self.m = 0
        self.b = 0
        self.params = None
        self.predicted_output = None
        self.train = None
        self.test = None
        
    
    def fit(self,dataset):
        ## The Model is designed such that it takes a dataframe as input containg only 2 Columns input & output
        
        #train test split using random shuffle
        
        elem = [i for i in range(dataset.shape[0])]
        random.shuffle(elem)
        train_size = (dataset.shape[0]*8)//10
        test_size = dataset.shape[0]-train_size
        self.train = dataset.iloc[elem[0:train_size]]
        self.test = dataset.iloc[elem[train_size:]]
        params = [] # Params is a list which will contain the value of m & b respectively
        output_train = [dataset.iloc[elem[0:90],1:].values][0]
        input_train = [dataset.iloc[elem[90:],:1].values][0]
        
        output_mean = 0
        input_mean = 0
        
        for  i in range(len(output)):
            output_mean = output_mean + output[i]
        output_mean = output_mean / len(output)
                
        for  i in input_:
            input_mean += i
        input_mean = input_mean/len(input_)
        
        y_error = []
        x_error = []
        
        for i in range(len(output)):
            y_error.append(np.round((output[i]-output_mean),2))
            x_error.append(np.round((input_[i]-input_mean),2))
        cov_xy = 0
        var_x = 0
        
        for i in range(len(y_error)):
            cov_xy +=  y_error[i]*x_error[i]
            var_x += x_error[i]*x_error[i]
        self.m = cov_xy/var_x
        self.b = output_mean - self.m*input_mean
        params.append(self.m)
        params.append(self.b)
        self.params = params
               
    def predict(self,test_input):
        output_predicted = []
        for i in range(len(test_input)):
            output_predicted.append(round((m*test_input[i] + b)[0],2))
        
        self.predicted_output = output_predicted
        return output_predicted
    
    def MSE(self,actul_input,input_predicted):
        self.MSE = 0
        for i in range(len(actul_input)):
            self.MSE += (actul_input[i]-input_predicted[i])**2
        self.MSE = self.MSE / len(actul_input)
        return self.MSE
        
        

In [40]:
a1 = SLR()

In [41]:
a1.fit(df)

In [42]:
print("Intercept Computed By Our Model: ",round(a1.params[1],5))
print("Intercept Computed By Sklearn Model: ",round(lr.intercept_[0],5))
print('\n')
print("Slope Computed by our Model: ",round(a1.params[0],5))
print("Slope COmputed by Sklearn Model: ",round(lr.coef_[0][0],5))

Intercept Computed By Our Model:  -9.38334
Intercept Computed By Sklearn Model:  -9.3853


Slope Computed by our Model:  2.16657
Slope COmputed by Sklearn Model:  2.1668


In [43]:
test_data = a1.test

In [44]:
test_data

CGPA   LPA
8     8.4   8.8
13    9.1  10.2
96    8.9   9.8
89    8.8   9.6
81    8.1   8.7
100   8.5   9.2
88    8.4   8.8
117   9.3  10.8
92    8.2   8.4
47    9.0  10.0
23    8.3   8.9
34    8.6   9.3
97    9.3  10.8
113   9.1  10.2
16    8.9   9.8
3     8.3   8.9
106   8.7   9.5
24    9.2  10.5
20    8.5   9.2
14    8.6   9.3
5     7.8   7.2
80    8.5   9.2
87    9.0  10.0
108   8.4   8.8

In [45]:
training_data = a1.train

In [46]:
predictions = a1.predict(training_data.CGPA.values.reshape(-1,1))

In [47]:
actul_values = training_data.LPA.values

In [48]:
a1.MSE(actul_input=actul_values,input_predicted=predictions)

0.031191666666666656

In [49]:
from sklearn.metrics import mean_absolute_error,mean_squared_error

In [50]:
mean_squared_error(actul_values,predictions)

0.031191666666666645